### PR TITLE
Use locking to checkout/read tree

### LINKED ISSUE #96: Race condition in Tree.node()
Try this
```python
import fmf
import threading

def checkout():
    fmf.base.Tree.node( {
            "url": "https://github.com/psss/fmf",
            "name": "/tests/docs"
        } ).name == '/tests/docs'

threads = []
for i in range(10):
    threads.append(threading.Thread(target=checkout))

for t in threads:
    t.start()

for t in threads:
    t.join()
```

And you will see a lot of errors:
1. `Command 'git clone https://github.com/psss/fmf /home/lzachar/.cache/fmf/https:__github.com_psss_fmf' returned non-zero exit status 128.` very likely to be 'directory exists and is non-empty'
2. `Command 'git checkout -f master' returned non-zero exit status 1.`

### PR DESCRIPTION
**Pull request by lukaszachy**
Todo:
1. Figure out timeout and retries (if repo is too large...)
2. What to do with stale lock and how to detect it?
3. Test for `Tree.node` as one needs to force git pull (otherwise it all happens too fast)

But please review the concept (resolves #96)
Two locks
1. To read tree (request for ref and read are atomic)
2. To fetch (clone /fetch and checkout are atomic)

### CODE REVIEW

File: fmf/utils.py [hunk 1]
```diff
@@ -38,6 +39,9 @@
 # Cache expiration in seconds
 CACHE_EXPIRATION = 1200
 
+# Lock timeout in seconds
+LOCK_TIMEOUT = 10
```
Comment (psss) [inline fmf/utils.py:43]: Sometime the fetching can take longer. I would extend this a bit, let's say 30 seconds?

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for implementing this. Looks good, added just a few comments. May one question: The `lockfile` docs mention that the package is deprecated. Shall we use [fasteners](https://github.com/harlowja/fasteners) or other project as suggested? Hm, but it seems it's not packaged for Python 2.

**Comment by psss** [comment]
> * Figure out timeout and retries (if repo is too large...)

For now I would just go with extended timeout and leave possible retry on the caller.

> * What to do with stale lock and how to detect it?

No idea about this. Briefly looked through the docs and they mentioned that symlinks are used for locking but haven't dived into details.

> * Test for `Tree.node` as one needs to force git pull (otherwise it all happens too fast)

Or, perhaps, a little bit larger tree would help? Because the main point here is to prevent branch switch when growing the metadata tree.

**Comment by lukaszachy** [comment]
> Thanks for implementing this. Looks good, added just a few comments. May one question: The `lockfile` docs mention that the Shall we use [fasteners](https://github.com/harlowja/fasteners) or other project as suggested? Hm, but it seems it's not packaged for Python 2.

I see it in epel-7. I've briefly checked features and they are more or less same, so I see no problem with changing it to fasterns.

File: fmf/utils.py [hunk 2]
```diff
@@ -523,17 +527,23 @@ def fetch(url, ref='master'):
     # Prepare the destination path and the cache directory
     cache = os.path.expanduser(os.environ.get('XDG_CACHE_HOME', '~/.cache'))
     fmf_cache = os.path.join(cache, 'fmf')
-    if not os.path.isdir(fmf_cache):
-        try:
-            os.makedirs(fmf_cache)
-        except OSError:
+    try:
+        os.makedirs(fmf_cache, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply fmf/utils.py:531]: Ouch. These Python 2 compatibility is a real pain :)

File: fmf/utils.py (hunk 1 above)
Comment (lukaszachy) [reply fmf/utils.py:43]: This is something to discuss. We know it can be even minutes :/ (large repos and slow network)

**Comment by lukaszachy** [comment]
> No idea about this. Briefly looked through the docs and they mentioned that symlinks are used for locking but haven't dived into details.

For this simple locking snip 
```python
from lockfile import LockFile
lock = LockFile('/tmp/zzz')
lock.acquire()
```
lockfile creates these files (hard linked)
```
ls -li /tmp/marvin-42-4fbf3740.502065385138845943249880 /tmp/zzz.lock 
1506 -rw-r--r--. 2 lzachar lzachar 5 Feb 10 18:49 /tmp/marvin-42-4fbf3740.502065385138845943249880
1506 -rw-r--r--. 2 lzachar lzachar 5 Feb 10 18:49 /tmp/zzz.lock
```

I'm thinking about writing `os.getpid()` into lockfile. So at least we will see which process got the lock (and possibly check whether is is still running)

Thank you for the update, I'll resolve your comments soon

**Comment by lukaszachy** [comment]
> > Thanks for implementing this. Looks good, added just a few comments. May one question: The `lockfile` docs mention that the Shall we use [fasteners](https://github.com/harlowja/fasteners) or other project as suggested? Hm, but it seems it's not packaged for Python 2.
> 
> I see it in epel-7. I've briefly checked features and they are more or less same, so I see no problem with changing it to fasterns.

I've looked into fasteners but they don't support timeout for with statement. (https://github.com/harlowja/fasteners/issues/51)
Which means we will need to implement it on our own or use try/finally to limit the chance to create state locks.

Is there any benefit with using fasteners?

**Comment by thrix** [comment]
hmm, results look weird, retrying again

**Comment by psss** [comment]
> Is there any benefit with using fasteners?

I don't know of any. Ok, let's stick with the `lockfile` for now if it covers our needs well.

**Comment by lukaszachy** [comment]
For the future 
- lockfile has a bit slow polling timeout - interval is 1/10 of the timeout and cannot be changed.
- fasteners can modify this interval but you either have thread or process safe lock. And we need both (thread for testsuite, process for most of real life usage). 
- filelock seems to work the best - fast polling and usable both in real life and in the test suite

**Review by psss — CHANGES_REQUESTED** [review]
Looks good. Just one idea for improvement (lock file location) and two questions:

 * Is it expected that the lock file remains on the filesystem after the process is finished? After the test suite is finished I still see the lock there.
 * The coverage dropped to 99% thanks to `except Timeout` lines. Do you think it would be possible to similate timeouts?

File: fmf/base.py [hunk 3]
```diff
@@ -557,23 +560,37 @@ def node(reference):
         # Fetch remote git repository
         if 'url' in reference:
             path = reference.get('path', '.').lstrip('/')
-            repository = utils.fetch(
-                reference.get('url'), reference.get('ref'))
-            root = os.path.join(repository, path)
+            # Lock for the intention to fetch/read git from URL to the cache
+            # Add .lock extension
+            lock_path = os.path.join(
+                '/tmp', reference["url"].replace('/', '_')) + '.lock'
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply fmf/base.py:566]: Ouch, of course  - need to have the lock close to the destination.

**Comment by lukaszachy** [comment]
> Looks good. Just one idea for improvement (lock file location) and two questions:
> 
>     * Is it expected that the lock file remains on the filesystem after the process is finished? After the test suite is finished I still see the lock there.

https://stackoverflow.com/questions/58098634/why-does-the-python-filelock-library-delete-lockfiles-on-windows-but-not-unix

I'm not sure we can fix that.
 
>     * The coverage dropped to 99% thanks to `except Timeout` lines. Do you think it would be possible to similate timeouts?

Will do

**Comment by psss** [comment]
> I'm not sure we can fix that.

I see, then this is just ok.

**Comment by psss** [comment]
Thanks for all the improvements and keeping the coverage at 100%!

**Comment by lukaszachy** [comment]
Having 100% coverage is really nice. Very good policy from the start - thank you for setting this qualify level

**Pull request merged**